stock_reserve_rule/models/stock_reserve_rule.py

```diff
@@ -0,0 +1,256 @@
+    removal_strategy = fields.Selection(
+        string="Advanced Removal Strategy",
+        selection=[
+            ("default", "Default Removal Strategy"),
+            ("empty_bin", "Empty Bins"),
+            ("packaging", "Prefer Full Packaging"),
+        ],
+        required=True,
+        default="default",
+    )
+
+    def _eval_quant_domain(self, quants, domain):
+        quant_domain = [("id", "in", quants.ids)]
+        return self.env["stock.quant"].search(
+            expression.AND([quant_domain, domain])
+        )
+
+    def _filter_quants(self, move, quants):
+        domain = safe_eval(self.quant_domain) or []
+        if domain:
+            return self._eval_quant_domain(quants, domain)
+        return quants
+
+    def _apply_strategy(self, quants):
+        """Apply the advanced removal strategy
+
+        New methods can be added by:
+
+        - Adding a selection in the 'removal_strategy' field.
+        - adding a method named after the selection value
+          (_apply_strategy_SELECTION)
+
+        A strategy has to comply with this signature: (self, quants)
+        Where 'self' is the current rule and 'quants' are the candidate
+        quants allowed for the rule, sorted by the company's removal
+        strategy (fifo, fefo, ...).
+        It has to get the initial need using 'need = yield' once, then,
+        each time the strategy decides to take quantities in a location,
+        it has to yield and retrieve the remaining needed using:
+
+            need = yield location, location_quantity, quantity_to_take
+
+        See '_apply_strategy_default' for a short example.
+
+        """
+        method_name = "_apply_strategy_%s" % (self.removal_strategy)
+        yield from getattr(self, method_name)(quants)
+
+    def _apply_strategy_default(self, quants):
+        need = yield
+        # Propose quants in the same order than returned originally by
+        # the _gather method, so based on fifo, fefo, ...
+        for quant in quants:
+            need = yield (
+                quant.location_id,
+                quant.quantity - quant.reserved_quantity,
+                need,
+            )
+
+    def _apply_strategy_empty_bin(self, quants):
+        need = yield
+        # Group by location (in this removal strategies, we want to consider
+        # the total quantity held in a location).
+        quants_per_bin = quants._group_by_location()
+
+        # Sort by min quant first so we empty as most bins as we
+        # can. But keep the original ordering for equal quantities!
+        bins = sorted(
+            [
+                (
+                    sum(quants.mapped("quantity"))
+                    - sum(quants.mapped("reserved_quantity")),
+                    quants,
+                    location,
+                )
+                for location, quants in quants_per_bin
+            ]
+        )
+
+        # Propose the smallest quants first, so we can empty as most bins
```

Comment:
This must be changed to choose the biggest quant (so invert the order). This makes less operational work. It is a bit stupid to start with the pallet with the few products on it and stack many products from other locations on it. You just want to start with the pallet with the most products and stack few additional products from other locations.